builder/builder.go

```diff
@@ -0,0 +1,228 @@
+package builder
+
+import (
+	"context"
+	"os"
+	"sort"
+
+	"github.com/docker/buildx/driver"
+	"github.com/docker/buildx/store"
+	"github.com/docker/buildx/store/storeutil"
+	"github.com/docker/buildx/util/imagetools"
+	"github.com/docker/buildx/util/progress"
+	"github.com/docker/cli/cli/command"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+// Builder represents an active builder object
+type Builder struct {
+	dockerCli       command.Cli
+	name            string
+	txn             *store.Txn
+	contextPathHash string
+
+	NodeGroup *store.NodeGroup
+	Drivers   []Driver
+	Err       error
+}
+
+// Option provides a variadic option for configuring the builder.
+type Option func(b *Builder) error
+
+// WithName sets builder name.
+func WithName(name string) Option {
+	return func(b *Builder) error {
+		b.name = name
+		return nil
+	}
+}
+
+// WithStore sets a store instance used at init.
+func WithStore(txn *store.Txn) Option {
+	return func(b *Builder) error {
+		b.txn = txn
+		return nil
+	}
+}
+
+// WithContextPathHash is used for determining pods in k8s driver instance.
+func WithContextPathHash(contextPathHash string) Option {
+	return func(b *Builder) error {
+		b.contextPathHash = contextPathHash
+		return nil
+	}
+}
+
+// New initializes a new builder client
+func New(dockerCli command.Cli, opts ...Option) (_ *Builder, err error) {
+	b := &Builder{
+		dockerCli: dockerCli,
+	}
+
+	for _, opt := range opts {
+		err := opt(b)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if b.txn == nil {
+		// if store instance is nil we create a short-lived one using the
+		// default store and ensure we release it on completion
+		var release func()
+		b.txn, release, err = storeutil.GetStore(dockerCli)
+		if err != nil {
+			return nil, err
+		}
+		defer release()
+	}
+
+	if b.name != "" {
+		if b.NodeGroup, err = storeutil.GetNodeGroup(b.txn, dockerCli, b.name); err != nil {
+			return nil, err
+		}
+	} else {
+		if b.NodeGroup, err = storeutil.GetCurrentInstance(b.txn, dockerCli); err != nil {
+			return nil, err
+		}
+	}
+	if b.NodeGroup.Name == "default" && len(b.NodeGroup.Nodes) == 1 {
+		b.NodeGroup.Name = b.NodeGroup.Nodes[0].Endpoint
+	}
+
+	return b, nil
+}
+
+// Validate validates builder context
+func (b *Builder) Validate() error {
```

Comment:
IMO, this should be called as part of `New` if possible.

Reply:
Yes in my initial attempt it was a client opt but then I thought it could be handy if we want this client to also create/update a builder. Will change that for now.

Reply:
Yeah, we can keep it as a separate method as well, I just think we should call it in `New` by default.